R/sparsevctrs.R

```diff
@@ -1,3 +1,79 @@
 is_sparse_matrix <- function(x) {
   methods::is(x, "sparseMatrix")
 }
+
+toggle_sparsity <- function(object, data) {
+  toggle_sparse <- "no"
+
+  if (allow_sparse(object$fit$actions$model$spec)) {
+    if ("recipe" %in% names(object$pre$actions)) {
+      est_sparsity <- recipes::.recipes_estimate_sparsity(
+        object$pre$actions$recipe$recipe
+      )
+    } else {
+      est_sparsity <- sparsevctrs::sparsity(data, 1000)
+    }
+
+    pred_log_fold <- pred_log_fold(
+      est_sparsity,
+      object$fit$actions$model$spec$engine,
+      nrow(data)
+    )
+    if (pred_log_fold > 0) {
+      toggle_sparse <- "yes"
+    }
+  }
+
+  object$pre$actions$recipe$recipe <- recipes::.recipes_toggle_sparse_args(
+    object$pre$actions$recipe$recipe,
+    choice = toggle_sparse
+  )
+  object
+}
+
+allow_sparse <- function(x) {
+  if (inherits(x, "model_fit")) {
+    x <- x$spec
+  }
+  res <- parsnip::get_from_env(paste0(class(x)[1], "_encoding"))
+  all(res$allow_sparse_x[res$engine == x$engine])
+}
+
+pred_log_fold <- function(sparsity, model, n_rows) {
+  if (is.null(model) || model == "ranger") {
+    return("no")
+  }
```

Comment:
<img width="647" alt="Screenshot 2025-01-15 at 1 32 52 PM" src="https://github.com/user-attachments/assets/9d41b06a-d91a-442a-813c-ac253b1af865" />

since `ranger` performs identically for both sparse and dense data, we turn off automatically